Improve the behavior of the custom @objc name in the available attribute when generates ObjC headers

pitiphong-p opened the pull request:
<!-- What's in this pull request? -->
This PR is an improvement of the #18081. It adds a few improvements for the custom `@objc` name in the available attribute when generates ObjC headers

Here's a list of the improvements
1. The compiler will follow the base type when the rename attribute references to the declaration in other Type
2. The compiler will try to look at the type of the method parameter when the rename attribute references to an overloading method

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

pitiphong-p committed:
Follow the base type if the renamed attribute is point to another type

pitiphong-p committed:
Find the renamed overloading method by looking at the type of method parameters

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -931,9 +931,23 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+      TypeDecl *typeDecl = declContext->getSelfNominalTypeDecl();
+      
+      if (!renamedDeclName.isSpecial()) {
+        auto contextIdentifier = D->getASTContext().getIdentifier(renamedParsedDeclName.ContextName);
```
Hm, what happens if there is no context name? (We use that for renames from one member of a type to another member of the same type.)

jrose-apple replied:
Also, why the check for special names?

pitiphong-p replied:
I forgot to handle the case of no context name, will update to handle it soon. On the special name, I found some case where the renamed is initializer and I think it causes an error in the `getIdentifier` method. Do you have any suggestion for handling this case?

jrose-apple replied:
Since you're calling `getIdentifier` on the context name rather than the member name, I don't think the test for `isSpecial` is doing anything useful. The call to `formDeclName` above should already handle that case.

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -944,10 +958,26 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+          bool hasSameParameterTypes = true;
+          for (auto index : indices(*cParams)) {
+            if (cParams->get(index)->getType()->getCanonicalType() !=
+                dParams->get(index)->getType()->getCanonicalType()) {
```
Hm…I think checking canonical types is probably too strict. What if someone changes optionality, or uses a superclass instead of a subclass? `TypeBase::matchesParameter` might be a good thing to look at here, probably using the `AllowOverride` and `AllowTopLevelOptionalMismatch` flags.

Another option would be to only worry about matching types if there are multiple overloads. That both saves work and offers a probably-better experience in the weird cases I mentioned above.

pitiphong-p replied:
Thank you for the suggestion, I'll update this soon

jrose-apple reviewed (commented):
Thanks for circling back to this. I think there's still some refinement to do here.

pitiphong-p committed:
Checking if the ContextName of the Renamed Decl is empty or not instead of checking if it’s a special decl instead

pitiphong-p committed:
Use the matchesParameter method instead of direct comparison to check the type of the parameter

pitiphong-p commented:
@jrose-apple I updated this PR with your suggestions. Please review it again.

jrose-apple commented:
Thanks, the new code looks good, but now you need test cases to try out the no-context and new overloading scenarios.

jrose-apple commented on test/PrintAsObjC/availability.swift:
```diff
@@ -195,17 +197,17 @@
+// CHECK-SAME: SWIFT_UNAVAILABLE_MSG("'unavailableMethodInUnavailableClass' has been renamed to 'replacingMethodInReplacementClassWithFirst:second:': use method in another class instead")
 // CHECK-NEXT: - (void)unavailableOnMacOSMethodInUnavailableClassWithFirst:(NSInteger)first second:(NSInteger)second
-// CHECK-SAME: SWIFT_AVAILABILITY(macos,unavailable,message="'unavailableOnMacOSMethodInUnavailableClass' has been renamed to 'ReplacementAvailableProtocol.methodReplacingInReplacementProtocol(first:second:)': use method in another class instead");
+// CHECK-SAME: SWIFT_AVAILABILITY(macos,unavailable,message="'unavailableOnMacOSMethodInUnavailableClass' has been renamed to 'replacingMethodInReplacementProtocolWithFirst:second:': use method in another class instead");
```
…I'm having second thoughts about this change. We lost the information that says where the new method is; it sounds like you just need to swap it out with the current method. Maybe it's better not to do the context thing.

jrose-apple commented:
Ah, I didn't read the test cases closely enough to see that there already were some renames without contexts. Except now I'm thinking the renames-with-contexts aren't necessarily better!

For the type-matching change, you should be able to do this:

```swift
class Foo {
  @objc(processAnything:)
  func processAnything(_: NSObject) {}
  @objc(processInteger:)
  func processAnything(_: Int) {} // just for confusion's sake

  @available(*, unavailable, renamed: "processAnything(_:)")
  @objc func processSpecificThing(_: UIView) {}
}
```

The types don't match up in this case, but the new function can handle everything the old function can. That's why the "override" check is a good one.

pitiphong-p commented:
@jrose-apple Yeah, I do have a concern on the renames-with-context too but my thought was that let say we want to rename the whole class and also its methods too. Could this renames-with-context can be useful for this case? Click fix-it for the type first and then its methods.

jrose-apple commented:
I think if you do rename an entire class and its methods, you'd want to put the renaming member decls on the *new* class, because all your instances would suddenly pick up the new class. It's only static members that could benefit from this.

pitiphong-p commented:
I got your point and agree with it too. Then should I revert it and have only the overloading scenaior?

jrose-apple commented:
It might still be worth checking for the context and exiting early if it's present, so you don't print the *wrong* name.

Sorry for having you do extra work!

pitiphong-p commented:
You mean fall back to the old behavior by printing the Swift name for this case?

jrose-apple commented:
Yeah, that seems better than printing something misleading to me.

pitiphong-p commented:
Got it. Will update this PR :)

pitiphong-p committed:
Revert part of the `Follow the base type` to follow only in its own superclasses

pitiphong-p commented:
@jrose-apple I updated the code as you suggested. However I tweaked the behavior from what we discussed by still follow the base (context name) type but only within its own superclasses hierarchy. Is that OK with you? Or should I remove the all of the following base name feature?

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -941,7 +944,23 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+          fprintf(stderr, "Declared type : %s\n", typeDecl->getDeclaredInterfaceType()->getStringAsComponent().data());
+          fprintf(stderr, "Looked up type : %s\n", specificTypeLookup.getSingleTypeResult()->getDeclaredInterfaceType()->getStringAsComponent().data());
+          fprintf(stderr, "Result %d\n", specificTypeLookup.getSingleTypeResult()->getDeclaredInterfaceType()->matches(typeDecl->getDeclaredInterfaceType(), matchMode));
+          fprintf(stderr, "Result %d\n", typeDecl->getDeclaredInterfaceType()->matches(specificTypeLookup.getSingleTypeResult()->getDeclaredInterfaceType(), matchMode));
```
Don't forget to take this debugging code back out!

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -941,7 +944,23 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+        }
+        if (!specificTypeLookup.getSingleTypeResult()) {
+          renamedDecl = nullptr;
+          goto printing_rename_attribute;
```
Can we break this out into multiple functions, so we don't need a `goto`? Something like `tryPrintObjCRenameForAvailableAttr`.

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -933,6 +933,9 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
       TypeDecl *typeDecl = declContext->getSelfNominalTypeDecl();
       
+      TypeMatchOptions matchMode = TypeMatchFlags::AllowOverride;
+      matchMode |= TypeMatchFlags::AllowTopLevelOptionalMismatch;
```
I don't think you need AllowTopLevelOptionalMismatch for comparing just the types.

jrose-apple reviewed (commented):
Sorry to leave you hanging for two weeks while I was on vacation. Sure, allowing subclasses seems like a reasonable compromise.

pitiphong-p committed:
Remove debugging print statements

pitiphong-p committed:
Remove unnecsessary  TypeMatchFlags::AllowTopLevelOptionalMismatch matching option

pitiphong-p committed:
Extract the finding the proper objc renamed Decl into getRenameDecl method

pitiphong-p commented:
@jrose-apple Thank you for your comments and I updated the code with your comments already. Please review it again.

PS. I hope you had a very great time and vacation :D

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -912,29 +912,46 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+          return nullptr;
+        }
+        
+        if (typeDecl->getDeclaredInterfaceType()->matches(specificTypeLookup.getSingleTypeResult()->getDeclaredInterfaceType(), TypeMatchFlags::AllowOverride)) {
```
Nitpick: 80-column limit (here and below). It's okay to break this into multiple expressions.

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -912,29 +912,46 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+  const ValueDecl *getRenameDecl(const ValueDecl *D, const ParsedDeclName renamedParsedDeclName) {
     auto declContext = D->getDeclContext();
-    const ValueDecl *renamedDecl = nullptr;
+    auto renamedDeclName = renamedParsedDeclName.formDeclName(D->getASTContext());
```
Nitpick: `D->getASTContext()` is used multiple times; it would probably be best to save a reference to it early on.

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -912,29 +912,46 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
-        declContext->getSelfNominalTypeDecl(),
-        renamedDeclName, NL_QualifiedDefault, lookupResults);
+                                   typeDecl,
+                                   renamedDeclName, NL_QualifiedDefault, lookupResults);
```
Weird indentation here. When in doubt, the `clang-format` tool should do the right thing. (It'll be in the LLVM build directory.)

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -944,10 +961,29 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+          bool hasSameParameterTypes = true;
+          for (auto index : indices(*cParams)) {
+            TypeMatchOptions matchMode = TypeMatchFlags::AllowOverride;
+            matchMode |= TypeMatchFlags::AllowTopLevelOptionalMismatch;
```
If we're going to use these flags here, we need tests for them, i.e. where the parameter types don't match exactly but do follow the override rules.

jrose-apple replied:
(Alternately, you could take this back out, since we only decided to add it late in the game.)

jrose-apple reviewed (commented):
Sorry for the delay! I think I marked the notification as read but didn't actually get back to it.

pitiphong-p committed:
Introduce an ASTContext local variable in the method

pitiphong-p committed:
Reduce the Type Match scope when comparing the method parameters

pitiphong-p committed:
Run clang-format on the Print Availability and its related methods

pitiphong-p commented:
@jrose-apple Thank you for your feedback and don't worry about the delay. I updated my PR as you suggested. Please have a look.

jrose-apple commented:
Hm, are you sure your clang-format picked up the .clang-format file in the root of the source directory? There are some things in there that definitely don't match LLVM style, like the indented switch cases.

As for the matching flags, even AllowOverride needs tests, where we accept a superclass or AnyObject instead of a subclass. All of the tests so far just have Int parameters.

pitiphong-p commented:
@jrose-apple About the first question, I'm not sure since I ran the `clang-format` command via a command line and specify a full path. Will try another way.
And I also will write more tests for your second point

pitiphong-p committed:
Add more test case for the case of renamed methods have a parameter which is a superclass of itself

pitiphong-p committed:
Run clang-format on the Print Availability and its related methods

pitiphong-p commented:
@jrose-apple Updated as you suggested. Please have a look again.

jrose-apple commented on test/PrintAsObjC/availability.swift:
```diff
@@ -570,27 +595,49 @@ extension Availability {
+    func methodReplacingInReplacementClassWithPrimitiveParameters(first: Int, second: Int) -> Void {}
+    
+    @objc(replacingMethodInReplacementClassWithSuperclassObjectParametersWithFirst:second:)
+    func methodReplacingInReplacementClassWithClassObjectParameters(first: SubclassOfReplacementAvailable, second: SubclassOfReplacementAvailable) -> Void {}
```
See, it's good to have the test case: this is backwards! You can rename from a method that takes "Subclass" to a method that takes "Base", but not the other way around (because some existing calls might break).

Alternately, maybe we shouldn't be matching types at all, since sometimes people *do* do incompatible renames. What do you think?

jrose-apple replied:
> I did the `reversed type` parameter test based on the Variance type system.

Ah, but parameters are contravariant, not covariant.

> On the point of not matching types, I agree with you not considering the point you raised. I think The matching type is then extended work of the overloading lookup improvement.
> 
> So should we remove the matching type for now and may consider to implement it when someone asks in the forum?

Yeah, let's go with that. Sorry for sending you down a dead end!

pitiphong-p replied:
> Ah, but parameters are contravariant, not covariant.
Correct, parameters are contravariant thus I implemented the test that the parameters of the renamed to methods have the super type of the parameters of the deprecated method. Or I may misunderstood it and implemented the wrong cases. Could you check this for me again if I implemented it wrong or not

> Yeah, let's go with that. Sorry for sending you down a dead end!
Ok. Let’s go with that and please don’t worry about sending me down a dead end. I also want to go with the best approach for Swift too :)

pitiphong-p committed:
Remove the parent/sub type lookup logic

pitiphong-p committed:
Check the function parameter type only for the rename to overloaded methods

pitiphong-p commented:
@jrose-apple I updated the code. This time I remove the Allow Override checking when finding the renamed Decl. Also I don't compare the type if there is only one candidate in the lookup result but still compare the parameter types if there are multiple candidates in the result in case of the overloaded method lookup.

Please have a look

jrose-apple commented:
(Sorry for the latest delay. I was off for US Thanksgiving week and have a bunch of other stuff coming up this week. I haven't forgotten!)

pitiphong-p commented:
@jrose-apple Thank you for the update 😁

jrose-apple reviewed (commented):
Okay, I'm back and I have more comments. Sorry this keeps dragging out.

pitiphong-p commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -912,62 +911,109 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+    auto renamedDeclName = renamedParsedDeclName.formDeclName(astContext);
+
     if (isa<ClassDecl>(D) || isa<ProtocolDecl>(D)) {
       UnqualifiedLookup lookup(renamedDeclName.getBaseIdentifier(),
```
Then how should I handle the renamed to a nested class scenario here?

jrose-apple replied:
I think that's sufficient: if the context is non-empty, just print it as is, like we do with the methods.

pitiphong-p replied:
What's the case when a class is renamed to `nested class` here? Would it fallback to print as it is in the `nested class` scenario? I think we can mark a nested class with `@objc`

jrose-apple replied:
You can, but the compiler may or may not actually print it! There's a bug where it'll only get printed if it's nested within another `@objc` type. I think the original intention was to not print it at all, but we probably can't change that at this point.

pitiphong-p replied:
So you mean we should skip for the nested case and go with the Swift name?

pitiphong-p replied:
I’m sorry by 

> There's a bug where it'll only get printed if it's nested within another @objc type. I think the original intention was to not print it at all, but we probably can't change that at this point.

I’m not sure what you mean by this sentence. What part that get printed? The deprecation warning for the nested class?

pitiphong-p replied:
I see. Then I think we have 2 options here
1. Check if the Outer calsss is @objc compatible or not. If yes then do the ObjC naming lookup
2. Just ignore this case at the moment until we got a request

Is there anything other option here? And what do you think on this case?

pitiphong-p commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -912,62 +911,109 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+        if (!specificTypeLookup.getSingleTypeResult()) {
+          return nullptr;
+        }
+      }
```
I think what I want to do here is to check if the ContextName is valid and can be found, if not then we'll return null immediately

jrose-apple replied:
Just because we can find it doesn't mean it's the same context, though, right? What if I rename `Int.magnitude` to `Double.binade`?

…I mean, hopefully no one does this anyway, and I don't think we're handling it properly in Swift's own fix-it, but it'd be worse to suggest the *wrong* thing, right?

pitiphong-p replied:
What I did here is to get a context name and search for a type of that context. So if it is renamed to `Int.magnitude` then this would find a type result of `Int` and then it will continue to do a printing

This is for validating the the context is OK to use not to compare to the current one.

So should I remove this?

jrose-apple replied:
I'm not sure. Do you think we should trust that a context is sensible if the user writes it, or not trust any contexts at all?

pitiphong-p replied:
That’s a hard question. My initial thought is that it would be great if we check the context here since Xcode doesn’t autocomplete the name in the rename attributes and I typed a name wrong for a few times.

pitiphong-p replied:
I see. So we should remove this check here?

pitiphong-p replied:
It’s pretty late in my tome zone but I want to make sure that I understand it correctly. 

So you’re suggesting that we don’t need to lookup about the information of Context for now, do I understand your point correctly?

jrose-apple replied:
I think I'm leaning towards:

```
if renamed decl is a type {
  if the rename string has a "context" part {
    give up and use the rename string as is
  } else {
    do UnqualifiedLookup (as you have here)
  }
} else {
  do qualified lookup (as you have here), whether or not there's context
}
```

* * *

I also noticed that the `isa<ClassDecl> || isa<ProtocolDecl>` should also include `isa<EnumDecl>` (which also deserves a test). At this point that could be a separate PR if you want, though.

pitiphong-p replied:
Got it. Thanks. I'll update the PR with your suggestion.

On the including of checking for enum type, I think I will do it in this PR though if you don't mind

pitiphong-p replied:
@jrose-apple I was working on the Enum case as you suggested and then I found that Swift doesn't print the availability declaration for enum yet.

Should we add it here or in another PR?

jrose-apple replied:
Oof. Let's make that another PR, then.

pitiphong-p committed:
Remove duplicated test case

pitiphong-p committed:
Ignore the Context Name in the rename declaration on most of the cases

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -774,11 +774,10 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+ 
+    /// Returns \c true if anything was printed.
+  bool printAvailability(const Decl *D, PrintLeadingSpace printLeadingSpace =
+                                            PrintLeadingSpace::Yes) {
```
Nitpick: I think the existing formatting (double-indented wrapping) is preferred over splitting on the `=`. (Ultimately it's whatever `clang-format` says, though.)

pitiphong-p replied:
This is what `clang-format` says. Do you want me to change?

jrose-apple replied:
Huh. Guess not, sorry for the noise.

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -957,17 +986,26 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+    assert(!AvAttr->Rename.empty());
+
+    const swift::ValueDecl *renamedDecl =
+        getRenameDecl(D, parseDeclName(AvAttr->Rename));
```
Nitpick: no need for `swift::` here.

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -957,17 +986,26 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
     }
   }
-
+    
```
Nitpick: extra whitespace?

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -944,10 +956,27 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
-          continue;
+        if (isa<FuncDecl>(candidate)) {
+          auto cParams = cast<FuncDecl>(candidate)->getParameters();
+          auto dParams = cast<FuncDecl>(D)->getParameters();
```
I hate to scope-creep now, but should this handle any AbstractFunctionDecl, which would also cover initializers? (And in theory getters and setters but people don't usually put renames on those individually.)

pitiphong-p replied:
@jrose-apple you mean should we change from FuncDecl to be AbstractFuncDecl? Did I understand your point correctly?

jrose-apple replied:
Yes, that's what I was suggesting!

pitiphong-p replied:
Got it. Will try it on tomorrow and also look into the log about build failure too

jrose-apple commented:
@swift-ci Please test

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -912,29 +911,42 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+    
+    const ValueDecl *renamedDecl = nullptr;
+    SmallVector<ValueDecl *, 4> lookupResults;
+    declContext->lookupQualified(typeDecl, renamedDeclName,
```
Oops, you probably want `declContext->getDeclaredInterfaceType` here.

pitiphong-p replied:
Oh thank you. I changed this line to `declContext->lookupQualified(typeDecl ->getDeclaredInterfaceType...` instead of `declContext->getDeclaredInterfaceType` as you suggested. Is that OK

jrose-apple replied:
The difference is whether you take generic parameters into account in a constrained extension, but for Objective-C that shouldn't matter, since you can't expose a constrained extension method to Objective-C.

pitiphong-p committed:
Fixed: Build failed due to the merging/rebasing

pitiphong-p committed:
Change the isa<FuncDecl> checking to be AbstractFunctionDecl instead

pitiphong-p commented:
@jrose-apple I updated the code with your latest suggestions. Please have a look and run the test again :)

jrose-apple commented:
Ah, the formatting comments are still relevant, and I'm afraid I've made more work for you by suggesting the AbstractFunctionDecl thing: we need a test with overloaded initializers now.

pitiphong-p committed:
Run `clang-format` on the PrintAsObjC.cpp file

pitiphong-p committed:
Added: Overloaded initializer test cases

pitiphong-p commented:
@jrose-apple I fixed the format (sorry I forgot that previously) and added some test cases on the overloaded initializers cases.

However I'm not sure why the generated header for the subclass has `SWIFT_UNAVAILABLE` on all of the inherited initializers. I looked in the PrintAsObjC.cpp and found this

>    // Swift designated initializers are Objective-C designated initializers.
>    // This will only be reached if the overridden initializer has the required access

But those initializers don't have required access

jrose-apple commented:
The "unavailables" are for non-inherited initializers; that's separate from something being a designated initializer or not. You're good there.

Okay, this looks great! I also spotted another issue for a later PR: when we need to print a message instead of using the "renamed" field of an attribute, we say "'init' has been renamed to 'initWithFoo:bar:'". We're using the Swift base name instead of the ObjC selector!

That's *definitely* separable from this one, though. :-)

@swift-ci Please test

jrose-apple commented:
Nitpick for the future: please try *not* to clang-format parts of the file you aren't changing already!

jrose-apple commented:
Hm. What's different on iOS simulator?

pitiphong-p commented:
```
UNRESOLVED: Swift(iphonesimulator-x86_64) :: PrintAsObjC/availability.swift (1 of 1)
******************** TEST 'Swift(iphonesimulator-x86_64) :: PrintAsObjC/availability.swift' FAILED ********************
Exception during script execution:
Traceback (most recent call last):
  File "/Users/pitiphong.p/Work/Swift-Repo/llvm/utils/lit/lit/run.py", line 202, in _execute_test_impl
    result = test.config.test_format.execute(test, lit_config)
  File "/Users/pitiphong.p/Work/Swift-Repo/swift/test/swift_test.py", line 63, in execute
    result = super(SwiftTest, self).execute(test, litConfig)
  File "/Users/pitiphong.p/Work/Swift-Repo/llvm/utils/lit/lit/formats/shtest.py", line 25, in execute
    self.execute_external)
  File "/Users/pitiphong.p/Work/Swift-Repo/llvm/utils/lit/lit/TestRunner.py", line 1569, in executeShTest
    script = applySubstitutions(script, substitutions)
  File "/Users/pitiphong.p/Work/Swift-Repo/llvm/utils/lit/lit/TestRunner.py", line 1259, in applySubstitutions
    return list(map(processLine, script))
  File "/Users/pitiphong.p/Work/Swift-Repo/llvm/utils/lit/lit/TestRunner.py", line 1253, in processLine
    ln = re.sub(a, b, ln)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/re.py", line 155, in sub
    return _compile(pattern, flags).sub(repl, string, count)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/re.py", line 286, in _subx
    template = _compile_repl(template, pattern)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/re.py", line 271, in _compile_repl
    p = sre_parse.parse_template(repl, pattern)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/sre_parse.py", line 762, in parse_template
    s = Tokenizer(source)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/sre_parse.py", line 193, in __init__
    self.__next()
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/sre_parse.py", line 195, in __next
    if self.index >= len(self.string):
TypeError: object of type 'NoneType' has no len()


********************
Testing Time: 0.11s
********************
Unresolved Tests (1):
    Swift(iphonesimulator-x86_64) :: PrintAsObjC/availability.swift

  Unresolved Tests   : 1
```

Here's an error showing in my console when I try to run it on the simulator. How could I fix this?

pitiphong-p commented:
```UNRESOLVED: Swift(iphonesimulator-x86_64) :: PrintAsObjC/availability.swift (1 of 1)
******************** TEST 'Swift(iphonesimulator-x86_64) :: PrintAsObjC/availability.swift' FAILED ********************
Exception during script execution:
Traceback (most recent call last):
  File "/Users/<RADACTED>/Swift/llvm/utils/lit/lit/run.py", line 202, in _execute_test_impl
    result = test.config.test_format.execute(test, lit_config)
  File "/Users/<RADACTED>/Swift/swift/test/swift_test.py", line 63, in execute
    result = super(SwiftTest, self).execute(test, litConfig)
  File "/Users/<RADACTED>/Swift/llvm/utils/lit/lit/formats/shtest.py", line 25, in execute
    self.execute_external)
  File "/Users/<RADACTED>/Swift/llvm/utils/lit/lit/TestRunner.py", line 1569, in executeShTest
    script = applySubstitutions(script, substitutions)
  File "/Users/<RADACTED>/Swift/llvm/utils/lit/lit/TestRunner.py", line 1259, in applySubstitutions
    return list(map(processLine, script))
  File "/Users/<RADACTED>/Swift/llvm/utils/lit/lit/TestRunner.py", line 1253, in processLine
    ln = re.sub(a, b, ln)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/re.py", line 155, in sub
    return _compile(pattern, flags).sub(repl, string, count)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/re.py", line 286, in _subx
    template = _compile_repl(template, pattern)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/re.py", line 271, in _compile_repl
    p = sre_parse.parse_template(repl, pattern)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/sre_parse.py", line 762, in parse_template
    s = Tokenizer(source)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/sre_parse.py", line 193, in __init__
    self.__next()
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/sre_parse.py", line 195, in __next
    if self.index >= len(self.string):
TypeError: object of type 'NoneType' has no len()


********************
Testing Time: 0.11s
********************
Unresolved Tests (1):
    Swift(iphonesimulator-x86_64) :: PrintAsObjC/availability.swift

  Unresolved Tests   : 1
```

Here's an error showing in my console when I try to run a test on simulator. How could I fix this?

jrose-apple commented:
Uh, hm. That sounds like one of the `%` substitutions isn't implemented correctly, rather than anything to do with this PR. I don't think that's what the error was last week, either.

@swift-ci Please test macOS

jrose-apple commented:
Just to check, you did build the iOS Simulator standard library, right? (I still wouldn't expect this error if you hadn't done that, but it's worth checking.)

jrose-apple commented:
```
/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/test/PrintAsObjC/availability.swift:182:16: error: CHECK-NEXT: is not on the line after the previous match
// CHECK-NEXT: @end
               ^
<stdin>:302:1: note: 'next' match was here
@end
^
<stdin>:300:116: note: previous match ended here
- (nonnull instancetype)initWithDeprecatedOnMacOSFirst:(NSInteger)first second:(NSInteger)second SWIFT_UNAVAILABLE;
                                                                                                                   ^
<stdin>:301:1: note: non-matching line after previous match is here
- (nonnull instancetype)initWithUnavailableOnMacOSFirst:(NSInteger)first second:(NSInteger)second SWIFT_UNAVAILABLE;
^
```

(when testing as iphonesimulator-i386)

pitiphong-p commented:
@jrose-apple How can I build `iOS Simulator standard library`? I did build with the given build script in the Swift repo

pitiphong-p commented:
My initial look founds that the Obj-C header printing will skip the unavailable inherited initializers in the subclass interface except in the iPhone Simulator platform. I don't know if this is a bug or intended behavior. If it's a bug then should I fix it in this PR and could you give me a hint or place to look at since I can't find the code for this print?

jrose-apple commented:
I do see the "unavailable on macOS" in the name, which means it's *not* unavailable on all platforms. I guess the simplest thing to do would be to weaken the CHECK-NEXT to a CHECK, allowing there to be extra content before the `@end`. What do you think?

pitiphong-p commented:
@jrose-apple That sounds great to me. I just learned that `CHECK` has a difference behavior than `CHECK-NEXT` I'll try with it and push ASAP

pitiphong-p committed:
Losen the test for the initizlier in the Subclass since it’s unavailable on some platforms and won’t be printed

pitiphong-p commented:
@jrose-apple I pushed with the change. Could you please run the test again?

Also what's about the crash that occurred on my machine? Should we fix it or left it there since it's only happens on my machine

jrose-apple commented:
I don't really know what to do about that. You can check your lib/swift/ directory to see if the iOS simulator stdlib was indeed built; if it wasn't, you may have to pass `--ios` explicitly to build-script. But yeah, if it's not blocking you from testing and you're the only one hitting it I don't know if it's worth spending time on.

@swift-ci Please test

jrose-apple commented:
Also you can take a look at https://llvm.org/docs/CommandGuide/FileCheck.html to find out everything FileCheck can do. :-)

pitiphong-p commented:
@jrose-apple thanks for the URL. I found that URL while I was looking for a documentation for those syntax. However I just skimmed and found a sentence saying that CHeCK-NEXT and CHECK have a similar behavior so I skipped reading on the detail. I think I was wrong hahaha

And I will try on the -iOS option you suggested and will update here when I got an update

jrose-apple commented:
Whew. Thank you again for for sticking with this PR so long!

pitiphong-p commented:
OMG OMG OMG Thank you for your help too 😄 🎉

pitiphong-p commented:
On the issue about a crash on my machine, it's like what you said. There's no content in the iphonesimulator folder in `lib/swift` folder. I tried to build with` --ios` option and run the test again. And IT WORKS!!!. I think we can leave this PR merged for good now :D
